False positive in array_init rule when using `map` that doesn't take a closure

### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Bug Report

```bash
$ Pods/SwiftLint/swiftlint lint --path file.swift                                         
Loading configuration from '.swiftlint.yml'
Linting Swift files at path file.swift
Linting 'file.swift' (1/1)
file.swift:1:17: warning: Array Init Violation: Prefer using Array(seq) than seq.map { $0 } to convert a sequence into an Array. (array_init)
Done linting! Found 1 violation, 0 serious in 1 file.
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? 0.23.0
* Installation method used (Homebrew, CocoaPods, building from source, etc)? CocoaPods
* Paste your configuration file:

```yml
opt_in_rules:
  - array_init
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
  If so, paste their relative paths and respective contents.

No

* Which Xcode version are you using (check `xcode-select -p`)?

Xcode 9

* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

```swift
foo.something { RouteMapper.map($0) }
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -10,7 +10,10 @@
 
 ##### Bug Fixes
 
-* None.
+* Fix false positive in `array_init` rule when using a `map` that
+  doesn't take a closure.  
+  [Marcelo Fabri](https://github.com/marcelofabri)
+  [#1878](https://github.com/realm/SwiftLint/issues/1878)
 
 ## 0.23.0: Permanent Press Cycle
 

--- a/Rules.md
+++ b/Rules.md
@@ -163,6 +163,11 @@ foo.map { $0! /* force unwrap */ }
 
 ```
 
+```swift
+foo.something { RouteMapper.map($0) }
+
+```
+
 </details>
 <details>
 <summary>Triggering Examples</summary>

--- a/Source/SwiftLintFramework/Rules/ArrayInitRule.swift
+++ b/Source/SwiftLintFramework/Rules/ArrayInitRule.swift
@@ -26,7 +26,8 @@ public struct ArrayInitRule: ASTRule, ConfigurationProviderRule, OptInRule {
             "foo.map { $0() }\n",
             "foo.map { ((), $0) }\n",
             "foo.map { $0! }\n",
-            "foo.map { $0! /* force unwrap */ }\n"
+            "foo.map { $0! /* force unwrap */ }\n",
+            "foo.something { RouteMapper.map($0) }\n"
         ],
         triggeringExamples: [
             "↓foo.map({ $0 })\n",
@@ -53,6 +54,8 @@ public struct ArrayInitRule: ASTRule, ConfigurationProviderRule, OptInRule {
         guard kind == .call, let name = dictionary.name, name.hasSuffix(".map"),
             let bodyOffset = dictionary.bodyOffset,
             let bodyLength = dictionary.bodyLength,
+            let nameOffset = dictionary.nameOffset,
+            let nameLength = dictionary.nameLength,
             let offset = dictionary.offset else {
                 return []
         }
@@ -66,7 +69,10 @@ public struct ArrayInitRule: ASTRule, ConfigurationProviderRule, OptInRule {
             return !SyntaxKind.commentKinds().contains(kind)
         }
 
-        guard isShortParameterStyleViolation(file: file, tokens: tokens) ||
+        guard let firstToken = tokens.first,
+            case let nameEndPosition = nameOffset + nameLength,
+            isClosureParameter(firstToken: firstToken, nameEndPosition: nameEndPosition, file: file),
+            isShortParameterStyleViolation(file: file, tokens: tokens) ||
             isParameterStyleViolation(file: file, dictionary: dictionary, tokens: tokens),
             let lastToken = tokens.last,
             case let bodyEndPosition = bodyOffset + bodyLength,
@@ -79,6 +85,20 @@ public struct ArrayInitRule: ASTRule, ConfigurationProviderRule, OptInRule {
                            severity: configuration.severity,
                            location: Location(file: file, byteOffset: offset))
         ]
+    }
+
+    private func isClosureParameter(firstToken: SyntaxToken,
+                                    nameEndPosition: Int,
+                                    file: File) -> Bool {
+        let length = firstToken.offset - nameEndPosition
+        guard length > 0,
+            case let contents = file.contents.bridge(),
+            let byteRange = contents.byteRangeToNSRange(start: nameEndPosition, length: length) else {
+                return false
+        }
+
+        let pattern = regex("\\A\\s*\\(?\\s*\\{")
+        return pattern.firstMatch(in: file.contents, options: .anchored, range: byteRange) != nil
     }
 
     private func containsTrailingContent(lastToken: SyntaxToken,

